### DISCUSSION TITLE
OpenGL interop: Need a way to purge VAO FBO cache.

### DISCUSSION THREAD

**Discussion by WangHoi** [DiligentGraphics/DiligentEngine#318] [Q&A]
When initializing the Engine by Attaching to Existing OpenGL/GLES Context.
And working with mutiple GL contexts.

We need a way to purge engine's internal VAO and FBO cache of one specific GL context,
when that context is about to be destroyed.

**Comment by TheMostDiligent** [comment]
But why do you need to explicitly purge it? The caches will be purged when the device is destroyed.

**Comment by WangHoi** [comment]
The OpenGL contexts recreated when window got resized，new float preview window opens, etc.
Recreate diligent RenderDevice is very slow, need to load and initialize resources again and again.
We are using shared OpenGL context to workaround it.
Current remain problem is VAO FBO cache memory leaking.

**Reply by TheMostDiligent** [reply]
[This change](https://github.com/DiligentGraphics/DiligentCore/commit/b3a2111eb080ac7c17a9a77c311c6d88bc948731) should allow purging the cache when updating the context.

**Reply by WangHoi** [reply]
Not working as expected, VERIFY() error I got:
"VAO cache is not empty. Are there any unreleased objects?"
"Only framebuffers without attachments can be left in the cache"

Looking through the code, VAO/FBOCache clear internal cache on release of PSO, IBuffer and ITexture.

But the vertex buffers, PSOs, texture attachmenet are shared between contexts, no release.

**Reply by WangHoi** [reply]
I also think add the PurgeCaches param to `DeviceContextGLImpl::UpdateCurrentGLContext` is wrong.
Because a new current OpenGL context is active now, and Diligent  may call glDeleteVertexArrays() in new context with old context's VAOs.

`RenderDeviceGLImpl::PurgeContextCaches` with old OpenGL context activated is the right way.

<hr />

The following steps works:
Comment out the VERIFY check in the dtors of VAOCache and FBOCache,
During context re-creation,  call `RenderDeviceGLImpl::PurgeContextCaches` before the old context is about to be released.

**Reply by WangHoi** [reply]
I'm experimenting with the following local modifications:

Added a `IDeviceContextGL::PurgeCurrentContextCaches()` .

```c++
// Implements `IDeviceContextGL::PurgeCurrentContextCaches()`
void DeviceContextGLImpl::PurgeCurrentContextCaches()
{
    auto NativeGLContext = m_pDevice->m_GLContext.GetCurrentNativeGLContext();
    if (NativeGLContext != NULL)
        m_pDevice->PurgeContextCaches(NativeGLContext);
}
```

And instead of commenting the VERIFY checks in VAO/FBOCache's destructors, add purge methods to them.

**Reply by TheMostDiligent** [reply]
Can you send the PR with your changes?

**Reply by WangHoi** [reply]
Sent a PR here:
https://github.com/DiligentGraphics/DiligentCore/pull/629

**Reply by TheMostDiligent** [reply]
Thank you! I will take a look

**Comment by TheMostDiligent** [comment]
I renamed `PurgeCurrentContextCaches` to `PurgeCurrentGLContextCaches` for consistency with `UpdateCurrentGLContext`

**Reply by WangHoi** [reply]
:smile: Thank you it works now.
